pkg/kab/uninstall_test.go

```diff
@@ -0,0 +1,244 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kab_test
+
+import (
+	"os"
+
+	"cnab-k8s-installer-base/pkg/apis/kab/v1alpha1"
+	"cnab-k8s-installer-base/pkg/client/clientset/versioned/fake"
+	"cnab-k8s-installer-base/pkg/kab"
+	"cnab-k8s-installer-base/pkg/kab/vendor_mocks"
+	"cnab-k8s-installer-base/pkg/kubectl/mocks"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/mock"
+	v12 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/testing"
+)
+
+var _ = Describe("LookupManifest Tests", func() {
+
+	var (
+		client         *kab.Client
+		mockCore       *vendor_mocks.CoreV1Interface
+		mockKubeClient *vendor_mocks.Interface
+		mockNamespace  *vendor_mocks.NamespaceInterface
+		fakeKabClient  *fake.Clientset
+		err            error
+	)
+
+	BeforeEach(func() {
+		mockKubeClient = new(vendor_mocks.Interface)
+		mockCore = new(vendor_mocks.CoreV1Interface)
+		mockNamespace = new(vendor_mocks.NamespaceInterface)
+		fakeKabClient = fake.NewSimpleClientset()
+		mockKubeClient.On("CoreV1").Return(mockCore)
+		mockCore.On("Namespaces").Return(mockNamespace)
+
+		client = kab.NewKnbClient(mockKubeClient, nil, fakeKabClient, nil, nil, nil)
+	})
+
+	Context("When there is error listing manifests", func() {
+		It("the error is returned to the caller", func() {
+			mockNamespace.On("List", mock.Anything).Return(nil, errors.NewUnauthorized("test error"))
+			_, err = client.LookupManifest("myInstallation")
+			Expect(err).To(MatchError("test error"))
+		})
+	})
+	Context("When the manifest does not exist in the cluster", func() {
+		It("throws an exception", func() {
+
+			mockNamespace.On("List", mock.Anything).Return(&v12.NamespaceList{
+				Items: []v12.Namespace{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "my-ns",
+						},
+					},
+				},
+			}, nil)
+			fakeKabClient.PrependReactor("get", "*", func(action testing.Action) (handled bool, ret runtime.Object, err error) {
+				return true, nil, errors.NewNotFound(schema.GroupResource{}, "none")
+			})
+			_, err = client.LookupManifest("myInstallation")
+			Expect(err).To(MatchError("could not find manifest for installation name: myInstallation"))
+		})
+	})
+	Context("When there are more than one namespaces", func() {
+		It("all namespaces are looked up to find a manifest", func() {
```

Comment:
What about the edge case where a manifest somehow exists in more than one namespace? Unless we can guarantee that this never happens (e.g. do we check during installation that the manifest isn't already present in another namespace?), we should test for it?

Reply:
Yes, installation fails if it finds a manifest.

Reply:
Great, thanks.